File: firestorm-java/src/main/java/com/raylabz/firestorm/firestore/FSPaginator.java
package com.raylabz.firestorm.firestore;

import com.google.api.core.ApiFuture;
import com.google.api.core.ApiFutures;
import com.google.cloud.firestore.*;
import com.raylabz.firestorm.Firestorm;
import com.raylabz.firestorm.api.Filterable;
import com.raylabz.firestorm.async.FSFuture;
import com.raylabz.firestorm.exception.FirestormException;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

/**
 * Enables easy pagination.
 * @param <T> The type of objects returned by the paginator.
 * @author Nicos Kasenides
 * @version 2.0.0
 */
public class FSPaginator<T> extends Filterable<Query, T> {

    private final int DEFAULT_LIMIT = 10;
    private final String lastDocumentID;
    private int limit = DEFAULT_LIMIT;

    /**
     * Instantiates a Paginator object for a certain class.
     *
     * @param objectClass The type of objects returned by the Paginator.
     */
    private FSPaginator(Class<T> objectClass, final String lastDocumentID) {
        super(FS.getFirestore().collection(objectClass.getSimpleName()), objectClass);
        this.lastDocumentID = lastDocumentID;
    }

    /**
     * Instantiates a Paginator object for a certain class and a certain results limit.
     *
     * @param objectClass The type of objects returned by the Paginator.
     * @param limit       The limit in number of results for each page.
     */
    private FSPaginator(Class<T> objectClass, final String lastDocumentID, int limit) {
        this(objectClass, lastDocumentID);
        this.limit = limit;
    }

    public static <T> FSPaginator<T> next(Class<T> objectClass, final String lastDocumentID, final int limit) {
        return new FSPaginator<>(objectClass, lastDocumentID, limit);
    }

    public static <T> FSPaginator<T> next(Class<T> objectClass, final String lastDocumentID) {
        return new FSPaginator<>(objectClass, lastDocumentID);
    }

    /**
     * Filters by value (equality).
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereEqualTo(@Nonnull String field, @Nullable Object value) {
        query = query.whereEqualTo(field, value);
        return this;
    }

    /**
     * Filters by value by field path (equality).
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereEqualTo(@Nonnull FieldPath fieldPath, @Nullable Object value) {
        query = query.whereEqualTo(fieldPath, value);
        return this;
    }

    /**
     * Filters by value (less than).
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereLessThan(@Nonnull String field, @Nonnull Object value) {
        query = query.whereLessThan(field, value);
        return this;
    }

    /**
     * Filters by value (less than).
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereLessThan(@Nonnull FieldPath fieldPath, @Nonnull Object value) {
        query = query.whereLessThan(fieldPath, value);
        return this;
    }

    /**
     * Filters by value (less than or equal to).
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereLessThanOrEqualTo(@Nonnull String field, @Nonnull Object value) {
        query = query.whereLessThanOrEqualTo(field, value);
        return this;
    }

    /**
     * Filters by value (less than or equal to).
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereLessThanOrEqualTo(@Nonnull FieldPath fieldPath, @Nonnull Object value) {
        query = query.whereLessThanOrEqualTo(fieldPath, value);
        return this;
    }

    /**
     * Filters by value (greater than).
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereGreaterThan(@Nonnull String field, @Nonnull Object value) {
        query = query.whereGreaterThan(field, value);
        return this;
    }

    /**
     * Filters by value (greater than).
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereGreaterThan(@Nonnull FieldPath fieldPath, @Nonnull Object value) {
        query = query.whereGreaterThan(fieldPath, value);
        return this;
    }

    /**
     * Filters by value (greater than or equal to).
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereGreaterThanOrEqualTo(@Nonnull String field, @Nonnull Object value) {
        query = query.whereGreaterThanOrEqualTo(field, value);
        return this;
    }

    /**
     * Filters by value (greater than or equal to).
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereGreaterThanOrEqualTo(@Nonnull FieldPath fieldPath, @Nonnull Object value) {
        query = query.whereGreaterThanOrEqualTo(fieldPath, value);
        return this;
    }

    /**
     * Filters by array field containing a value.
     *
     * @param field The field.
     * @param value The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereArrayContains(@Nonnull String field, @Nonnull Object value) {
        query = query.whereArrayContains(field, value);
        return this;
    }

    /**
     * Filters by array field containing a value.
     *
     * @param fieldPath The field path.
     * @param value     The value.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereArrayContains(@Nonnull FieldPath fieldPath, @Nonnull Object value) {
        query = query.whereArrayContains(fieldPath, value);
        return this;
    }

    /**
     * Filters by field containing any of a list of values.
     *
     * @param field  The field.
     * @param values The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereArrayContainsAny(@Nonnull String field, @Nonnull List<?> values) {
        query = query.whereArrayContainsAny(field, values);
        return this;
    }

    /**
     * Filters by field containing any of a list of values.
     *
     * @param fieldPath The field path.
     * @param values    The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereArrayContainsAny(@Nonnull FieldPath fieldPath, @Nonnull List<?> values) {
        query = query.whereArrayContainsAny(fieldPath, values);
        return this;
    }

    /**
     * Filters by an <u>array</u> field containing a list of values.
     *
     * @param field  The field.
     * @param values The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereIn(@Nonnull String field, @Nonnull List<?> values) {
        query = query.whereIn(field, values);
        return this;
    }

    /**
     * Filters by an <u>array</u> field containing a list of values.
     *
     * @param fieldPath The field path.
     * @param values    The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereIn(@Nonnull FieldPath fieldPath, @Nonnull List<?> values) {
        query = query.whereIn(fieldPath, values);
        return this;
    }

    /**
     * Filters by an <u>array</u> field NOT containing a list of values.
     *
     * @param field  The field.
     * @param values The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereNotIn(@Nonnull String field, @Nonnull List<?> values) {
        query = query.whereNotIn(field, values);
        return this;
    }

    /**
     * Filters by an <u>array</u> field NOT containing a list of values.
     *
     * @param fieldPath The field path.
     * @param values    The list of values.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> whereNotIn(@Nonnull FieldPath fieldPath, @Nonnull List<?> values) {
        query = query.whereNotIn(fieldPath, values);
        return this;
    }

    /**
     * Orders results by a field.
     *
     * @param field The field.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> orderBy(@Nonnull String field) {
        query = query.orderBy(field);
        return this;
    }

    /**
     * Orders results by a field.
     *
     * @param fieldPath The field path.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> orderBy(@Nonnull FieldPath fieldPath) {
        query = query.orderBy(fieldPath);
        return this;
    }

    /**
     * Orders results by a field in a specified direction.
     *
     * @param field The field.
     * @param direction  The direction (ascending/descending) of ordering.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> orderBy(@Nonnull String field, @Nonnull Query.Direction direction) {
        query = query.orderBy(field, direction);
        return this;
    }

    /**
     * Orders results by a field in a specified direction.
     *
     * @param fieldPath The field path.
     * @param direction The direction (ascending/descending) of ordering.
     * @return Returns a Paginator.
     */
    @Nonnull
    public FSPaginator<T> orderBy(@Nonnull FieldPath fieldPath, @Nonnull Query.Direction direction) {
        query = query.orderBy(fieldPath, direction);
        return this;
    }

    /**
     * Fetches the result.
     * @return Returns an {@link FSFuture}.
     */
    public FSFuture<FSQueryResult<T>> fetch() {

        //If there is a last document, set the query to start after it:
        if (lastDocumentID != null) {
            DocumentReference lastDocumentReference = FS.getFirestore().collection(objectClass.getSimpleName()).document(lastDocumentID);
            try {
                DocumentSnapshot lastDocumentSnapshot = lastDocumentReference.get().get();
                if (lastDocumentSnapshot != null) {
                    query = query.startAfter(lastDocumentSnapshot);
                }
            } catch (ExecutionException | InterruptedException e) {
                throw new FirestormException(e);
            }
        }

        //Query limits:
        query = query.limit(limit);

        //Run the query and return the results:
        ApiFuture<QuerySnapshot> nextResultsFuture = query.get();
        ApiFuture<FSQueryResult<T>> resultsFuture = ApiFutures.transform(nextResultsFuture, nextResultsSnapshot -> {
            List<QueryDocumentSnapshot> documents = nextResultsSnapshot.getDocuments();
            ArrayList<T> documentList = new ArrayList<>();
            String lastID = null;
            for (final QueryDocumentSnapshot document : documents) {
                T object = document.toObject(objectClass);
                documentList.add(object);
                lastID = document.getId();
            }
            if (documentList.isEmpty()) {
                return new FSQueryResult<>(documentList, null, null);
            }
            else {
                return new FSQueryResult<>(documentList, documents.get(documents.size() - 1), lastID);
            }
        }, Firestorm.getSelectedExecutor());
        return FSFuture.fromAPIFuture(resultsFuture);
    }

}

File: firestorm_flutter/lib/fs/queries/fs_paginator.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firestorm/fs/queries/fs_query_result.dart';
import '../../firestorm.dart';
import '../fs.dart';

/// A paginator for Firestore queries that allows fetching objects in pages.
class FSPaginator<T> {

  late String? lastDocumentID;
  final String? subcollection;
  int numOfDocuments;
  Query query;

  /// Creates a paginator for Firestore queries.
  FSPaginator({ this.lastDocumentID, this.numOfDocuments = 10, this.subcollection }) : query = FS.reference.collection(T, subcollection: subcollection) {
    query = query.orderBy("id");
  }

  /// Filters results by equality.
  FSPaginator<T> whereEqualTo(String field, dynamic value) {
    query = query.where(field, isEqualTo: value);
    return this;
  }

  /// Filters results by inequality.
  FSPaginator<T> whereNotEqualTo(String field, dynamic value) {
    query = query.where(field, isNotEqualTo: value);
    return this;
  }

  /// Filters results by value (less than).
  FSPaginator<T> whereLessThan(String field, dynamic value) {
    query = query.where(field, isLessThan: value);
    return this;
  }

  /// Filters results by value (less than or equal to).
  FSPaginator<T> whereLessThanOrEqualTo(String field, dynamic value) {
    query = query.where(field, isLessThanOrEqualTo: value);
    return this;
  }

  /// Filters results by value (greater than).
  FSPaginator<T> whereGreaterThan(String field, dynamic value) {
    query = query.where(field, isGreaterThan: value);
    return this;
  }

  /// Filters results by value (greater than or equal to).
  FSPaginator<T> whereGreaterThanOrEqualTo(String field, dynamic value) {
    query = query.where(field, isGreaterThanOrEqualTo: value);
    return this;
  }

  /// Filters results by array field containing a value.
  FSPaginator<T> whereArrayContains(String field, dynamic value) {
    query = query.where(field, arrayContains: value);
    return this;
  }

  /// Filters results by array field containing a value.
  FSPaginator<T> whereArrayContainsAny(String field, List<dynamic> values) {
    query = query.where(field, arrayContainsAny: values);
    return this;
  }

  /// Filters results by array field containing any of the provided values.
  FSPaginator<T> whereIn(String field, List<dynamic> values) {
    query = query.where(field, whereIn: values);
    return this;
  }

  /// Filters by an array field NOT containing any of the provided values.
  FSPaginator<T> whereNotIn(String field, List<dynamic> values) {
    query = query.where(field, whereNotIn: values);
    return this;
  }

  /// Orders the query results by a field's value.
  FSPaginator<T> orderBy(String field, { bool descending = false }) {
    query = query.orderBy(field, descending: descending);
    return this;
  }

  /// Limits the number of results.
  FSPaginator<T> limit(int limit) {
    query = query.limit(limit);
    return this;
  }

  /// Starts the query at a document where the field values provided are met.
  FSPaginator<T> startAt(List<dynamic> fieldValues) {
    query = query.startAt(fieldValues);
    return this;
  }

  /// Starts the query after a specific document where the field values provided are met.
  FSPaginator<T> startAfter(List<dynamic> fieldValues) {
    query = query.startAfter(fieldValues);
    return this;
  }

  /// Ends the query before a specific document where the field values provided are met.
  FSPaginator<T> endBefore(List<dynamic> fieldValues) {
    query = query.endBefore(fieldValues);
    return this;
  }

  /// Ends the query at a specific document where the field values provided are met.
  FSPaginator<T> endAt(List<dynamic> fieldValues) {
    query = query.endAt(fieldValues);
    return this;
  }

  /// Streams the results of the query as a stream of type T.
  Stream<T> stream() async* {
    Deserializer? deserializer = FS.deserializers[T];
    if (deserializer == null) {
      throw UnsupportedError('No serializer found for type: $T. Consider re-generating Firestorm data classes.');
    }
    Stream<QuerySnapshot<Object?>> snapshots = query.snapshots();
    await for (QuerySnapshot<Object?> snapshot in snapshots) {
      for (QueryDocumentSnapshot<Object?> doc in snapshot.docs) {
        yield deserializer(doc.data() as Map<String, dynamic>);
      }
    }
  }

  /// Fetches the next page of objects.
  Future<FSQueryResult<T>> next() async {
    Deserializer? deserializer = FS.deserializers[T];
    final String? className = FS.classNames[T];
    if (deserializer == null || className == null) {
      throw UnsupportedError('No deserializer/class name found for type: $T. Consider re-generating Firestorm data classes.');
    }

    //If there is a last document ID, start from after that
    if (lastDocumentID != null) {
      dynamic object = await FS.get.one<T>(lastDocumentID!, subcollection: subcollection);
      if (object != null) {
        if (object.id != null) {
          query = query.startAfter([object.id]);
        }
        else {
          throw ArgumentError('FSPaginator: The last document ID provided is not valid: $lastDocumentID');
        }
      }
    }

    //Limit the number of documents to fetch
    query = query.limit(numOfDocuments);

    //Run the query and return results:
    var documents = await query.get();
    List<T> objects = [];

    for (QueryDocumentSnapshot doc in documents.docs) {
      T object = deserializer(doc.data() as Map<String, dynamic>);
      objects.add(object);
    }
    if (objects.isEmpty) {
      return FSQueryResult(objects, null, null);
    }
    else {
      lastDocumentID = documents.docs[documents.docs.length - 1].id;
      return FSQueryResult(objects, documents.docs[documents.docs.length - 1], lastDocumentID);
    }
  }


}